File: codes/JavaScript/variable-change-observer.js
// Observe changes to a variable

class Observable {
	constructor(value) {
		this.function = () => {};
		this._value = value;
		this.oldValue = null;
	}

	set value(value) {
		console.log(value);
		this.oldValue = this._value;
		this._value = value;
		this.function();
	}

	get value() {
		return this._value;
	}

	onChange(func) {
		this.function = func;
	}
}
const watcher = new Observable(10);

watcher.onChange(() => {
	console.log(`I changed from ${watcher.oldValue} to ${watcher.value}`);
});

watcher.value = 23;

File: codes/Python/variable-change-observer.py
class Observable():
    def __init__(self, value):
        self.onChangeFunction = lambda: None
        self.onWriteFunction = lambda: None
        self._value = value
        self.oldValue = None

    def get_value(self):
        return self._value

    def set_value(self, value):
        self.oldValue = self._value
        self._value = value
        self.onWriteFunction()
        if not self.oldValue == self._value:
            self.onChangeFunction()

    def onChange(self, func):
        self.onChangeFunction = func

    def onWrite(self, func):
        self.onWriteFunction = func

    value = property(get_value, set_value)

OBSERVED_VARIABLE = Observable(20)

OBSERVED_VARIABLE.onChange(lambda: print(f"Changed value from {OBSERVED_VARIABLE.oldValue} to {OBSERVED_VARIABLE.value}"))
OBSERVED_VARIABLE.onWrite(lambda: print(f"Wrote {OBSERVED_VARIABLE.value}"))

OBSERVED_VARIABLE.value = 18 #Logs "Wrote 18" and "Changed value from 20 to 18"
OBSERVED_VARIABLE.value = 18 #Only logs "Wrote 18" since the value didn't change